README.md
INTRODUCTION
============
The buggalo script can collect various information about an
exception in a Python script as well as information about the
users system, such as XBMC and Python versions.

The collected information is then posted to the internet at a
predefined URL where the addon author can investigate the exception.

The script is somewhat similar to posting the xbmc.log to pastebin,
but is more specialised and doesn't contain superfluous information.
It is also better integrated into the user experience, the user only
has to decide if they want to submit the bug report or not.

The user will see a dialog as seen in this screenshot:
http://tommy.winther.nu/files/2011/12/script_error.png

HOW TO USE
==========
To use this script you must do these things besides importing it.

1.  Set buggalo.SUBMIT_URL to a full URL where the collected data is submitted.

2.  Surround the code you want to be covered by this script in a try..except block, such as:

    ```python
    try
        # addon logic
    except Exception:
        buggalo.onExceptionRaised()
    ```

    For plugin type addons, it is a good idea to include pretty much
    everything inside the try..except block.
    See this link for an example:
    https://github.com/xbmc-danish-addons/plugin.video.news.tv2.dk/blob/master/addon.py#L124

    For script type addons, besides the rule above, each event in
    your UI should include the try..except block as well.
    See this link for an example:
    https://github.com/twinther/script.tvguide/blob/master/gui.py#L140

    Optionally, instead of writing the try..except block yourself, you can decorate the function
    with @buggalo_try_except(). Function specific may be provide in the decorator:

    ```python
    @buggalo_try_except({'class' : 'MyWindowXML', 'method' : 'onInit', 'other_key' : 'other_value'})
    def onInit(self)
        pass
    ```

3.  Finally you must setup the website where the error report is submitted.
    A good starting point is my buggalo-web module on github:
    https://github.com/twinther/buggalo-web
    If you want to roll your own custom setup then take a look at the submit.php
    file which store the error report in the database.
    https://github.com/twinther/buggalo-web/blob/master/submit.php

    If you don't want to or can't setup your own website you can use the shared site at:
    http://buggalo.xbmc.info/
    In this case buggalo.SUBMIT_URL must be set to:
    http://buggalo.xbmc.info/submit.php

WHAT IS COLLECTED
=================
Five groups of information is collected beyond basic information
such as date and time.

*  System information
   OS name and version, kernel version, etc.
*  Addon information
   Addon id, name, version, path, etc.
*  XBMC Information
   Build version and date, the current skin and language
*  Execution information
   Python version and sys.argv
*  Exception information
   Type of exception, message and full stack trace

For further details take a look at the code in buggalo.py

---------------------------------------------------------------------

The latest code is always available at github:
https://github.com/twinther/script.module.buggalo

The module is named after a creature in my favorite animated show:
http://theinfosphere.org/Where_the_Buggalo_Roam

---------------------------------------------------------------------
2012.03.16 - twinther


changelog.txt
[B]Version 1.1.3 - 2013-05-18[/B]
- Added support for sending error reports to a gmail account
- Updated language files from Transifex

[B]Version 1.1.2 - 2012-09-09[/B]
- Improved UI compatibility with skins other than Confluence
- Updated movie quotes

[B]Version 1.1.1 - 2012-09-09[/B]
- Reworked error dialog layout a bit

[B]Version 1.1.0 - 2012-07-11[/B]
- Added custom error dialog with details view
- Added support for tracking userflow
- More robust handling of extra data

[B]Version 1.0.1 - 2012-03-16[/B]
- Added option to pass extra data along in the error report
  either through addExtraData(key, value) or directly in
  onExceptionRaised(data) as a dict or string.
- Added @buggalo_try_except function decorator to easily wrap
  a function in a try..except clause.

[B]Version 1.0.0 - 2012-02-04[/B]
- Initial module version
- Named after a creature in my favorite animated show:
  http://theinfosphere.org/Where_the_Buggalo_Roam


lib/gui.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
#
import client

import xbmc
import xbmcaddon
import xbmcgui

buggaloAddon = xbmcaddon.Addon(id = 'script.module.buggalo')

ACTION_PARENT_DIR = 9
ACTION_PREVIOUS_MENU = 10

class BuggaloDialog(xbmcgui.WindowXMLDialog):
    THANKS_YOU_VISIBLE_LABEL = 98
    DETAILS_VISIBLE_LABEL = 99
    CLOSE_BUTTON = 100
    SUBMIT_BUTTON = 101
    DETAILS_BUTTON = 102
    DETAILS_LIST = 103
    HEADING_LABEL = 110

    ERROR_MESSAGE_GROUP = 200
    DETAILS_GROUP = 201
    THANK_YOU_GROUP = 202

    def __new__(cls, serviceUrl, heading, data):
        return super(BuggaloDialog, cls).__new__(cls, 'buggalo-dialog.xml', buggaloAddon.getAddonInfo('path'))

    def __init__(self, serviceUrl, heading, data):
        super(BuggaloDialog, self).__init__()
        self.serviceUrl = serviceUrl
        self.heading = heading
        self.data = data
        self.detailsVisible = False

    def onInit(self):
        self.getControl(self.HEADING_LABEL).setLabel(self.heading)
        self.getControl(self.DETAILS_VISIBLE_LABEL).setVisible(not self.detailsVisible)
        self.getControl(self.THANKS_YOU_VISIBLE_LABEL).setVisible(True)
        listControl = self.getControl(self.DETAILS_LIST)

        try:
            for group in sorted(self.data.keys()):
                values = self.data[group]
                if type(values) == dict:
                    item = xbmcgui.ListItem(label = '[B]%s[/B]' % group)
                    listControl.addItem(item)
                    for key in values:
                            item = xbmcgui.ListItem(label = '    %s' % key, label2 = str(values[key]))
                            listControl.addItem(item)

                else:
                    item = xbmcgui.ListItem(label = '[B]%s[/B]' % group, label2 = str(values))
                    listControl.addItem(item)
        except Exception:
            item = xbmcgui.ListItem(label = buggaloAddon.getLocalizedString(91007))
            listControl.addItem(item)


    def onAction(self, action):
        if action.getId() in [ACTION_PARENT_DIR, ACTION_PREVIOUS_MENU]:
            self.close()

    def onClick(self, controlId):
        if controlId == self.CLOSE_BUTTON:
            self.close()

        elif controlId == self.SUBMIT_BUTTON:
            self.getControl(self.DETAILS_VISIBLE_LABEL).setVisible(True)
            self.getControl(self.THANKS_YOU_VISIBLE_LABEL).setVisible(False)

            client.submitData(self.serviceUrl, self.data)

            xbmc.sleep(2000)
            self.close()

        elif controlId == self.DETAILS_BUTTON:
            self.detailsVisible = not self.detailsVisible
            self.getControl(self.DETAILS_VISIBLE_LABEL).setVisible(not self.detailsVisible)


    def onFocus(self, control):
        pass

lib/client.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
#
import datetime
import os
import platform
import simplejson
import sys
import traceback
import urllib2

import xbmc
import xbmcaddon

def gatherData(type, value, tracebackInfo, extraData, globalExtraData):
    data = dict()
    data['version'] = 3
    data['timestamp'] = datetime.datetime.now().isoformat()

    system = dict()
    try:
        if hasattr(os, 'uname'):
            # Works on recent unix flavors
            (sysname, nodename, release, version, machine) = os.uname()
        else:
            # Works on Windows (and others?)
            (sysname, nodename, release, version, machine, processor) = platform.uname()

        system['nodename'] = nodename
        system['sysname'] = sysname
        system['release'] = release
        system['version'] = version
        system['machine'] = machine
    except Exception, ex:
        system['sysname'] = sys.platform
        system['exception'] = str(ex)
    data['system'] = system

    addon = xbmcaddon.Addon()
    addonInfo = dict()
    addonInfo['id'] = addon.getAddonInfo('id')
    addonInfo['name'] = addon.getAddonInfo('name')
    addonInfo['version'] = addon.getAddonInfo('version')
    addonInfo['path'] = addon.getAddonInfo('path')
    addonInfo['profile'] = addon.getAddonInfo('profile')
    data['addon'] = addonInfo

    xbmcInfo = dict()
    xbmcInfo['buildVersion'] = xbmc.getInfoLabel('System.BuildVersion')
    xbmcInfo['buildDate'] = xbmc.getInfoLabel('System.BuildDate')
    xbmcInfo['skin'] = xbmc.getSkinDir()
    xbmcInfo['language'] = xbmc.getInfoLabel('System.Language')
    data['xbmc'] = xbmcInfo

    execution = dict()
    execution['python'] = sys.version
    execution['sys.argv'] = sys.argv
    data['execution'] = execution

    exception = dict()
    exception['type'] = str(type)
    exception['value'] = str(value)
    exception['stacktrace'] = traceback.format_tb(tracebackInfo)
    data['exception'] = exception

    extraDataInfo = dict()
    try:
        for (key, value) in globalExtraData.items():
            extraDataInfo[key] = str(value)

        if type(extraData) == dict:
            for (key, value) in extraData.items():
                extraDataInfo[key] = str(value)
        elif extraData is not None:
            extraDataInfo[''] = str(extraData)
    except Exception, ex:
        extraDataInfo['exception'] = str(ex)
    data['extraData'] = extraDataInfo

    return data


def submitData(serviceUrl, data):
    for attempt in range(0, 3):
        try:
            json = simplejson.dumps(data)
            req = urllib2.Request(serviceUrl, json)
            req.add_header('Content-Type', 'text/json')
            u = urllib2.urlopen(req)
            u.read()
            u.close()
            break # success; no further attempts
        except Exception:
            pass # probably timeout; retry




lib/buggalo.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
#
import sys
import traceback as tb
import random

import xbmcaddon
import xbmcplugin

import buggalo_client as client
import buggalo_gui as gui
import buggalo_userflow as userflow

# You must provide either the SUBMIT_URL or GMAIL_RECIPIENT
# via buggalo.SUBMIT_URL = '' or buggalo.GMAIL_RECIPIENT = ''

# The full URL to where the gathered data should be posted.
SUBMIT_URL = None
# The email address where the gathered data should be sent.
GMAIL_RECIPIENT = None

EXTRA_DATA = dict()

SCRIPT_ADDON = len(sys.argv) == 1

if not SCRIPT_ADDON:
    # Automatically track userflow for plugin type addons
    userflow.trackUserFlow('%s%s' % (sys.argv[0], sys.argv[2]))

def addExtraData(key, value):
    EXTRA_DATA[key] = value

def trackUserFlow(value):
    """
    Registers an entry in the user's flow through the addon.
    The values is stored in a dict with the current time as key and the provided value as the value.

    For plugin-type addons the user flow is automatically registered for each page the user loads.
    The value can be any string, so it's also useful in script-type addons.

    @param value: the value indicating the user's flow.
    @type value: str
    """
    userflow.trackUserFlow(value)

def getRandomHeading():
    """
    Get a random heading for use in dialogs, etc.
    The heading contains a random movie quote from the English strings.xml
    """
    return getLocalizedString(random.randint(90000, 90011))


def getLocalizedString(id):
    """
    Same as Addon.getLocalizedString() but retrieves data from this module's strings.xml
    """
    buggaloAddon = xbmcaddon.Addon(id = 'script.module.buggalo')
    return buggaloAddon.getLocalizedString(id)


def buggalo_try_except(extraData = None):
    """
    @buggalo_try_except function decorator wraps a function in a try..except clause and invokes onExceptionRaised()
    in case an exception is raised. Provide extraData to specific function specific extraData.

    @param extraData: str or dict
    """
    def decorator(fn):
        def wrap_in_try_except(*args, **kwargs):
            try:
                return fn(*args, **kwargs)
            except Exception:
                onExceptionRaised(extraData)
        return wrap_in_try_except
    return decorator

def onExceptionRaised(extraData = None):
    """
    Invoke this method in an except clause to allow the user to submit
    a bug report with stacktrace, system information, etc.

    This also avoids the 'Script error' popup in XBMC, unless of course
    an exception is thrown in this code :-)

    @param extraData: str or dict
    """
    # start by logging the usual info to stderr
    (type, value, traceback) = sys.exc_info()
    tb.print_exception(type, value, traceback)

    if not SCRIPT_ADDON:
        try:
            # signal error to XBMC to hide progress dialog
            HANDLE = int(sys.argv[1])
            xbmcplugin.endOfDirectory(HANDLE, succeeded=False)
        except Exception:
            pass

    heading = getRandomHeading()
    data = client.gatherData(type, value, traceback, extraData, EXTRA_DATA)

    d = gui.BuggaloDialog(SUBMIT_URL, GMAIL_RECIPIENT, heading, data)
    d.doModal()
    del d


lib/buggalo_gui.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
#
import buggalo_client as client

import xbmc
import xbmcaddon
import xbmcgui

buggaloAddon = xbmcaddon.Addon(id = 'script.module.buggalo')

ACTION_PARENT_DIR = 9
ACTION_PREVIOUS_MENU = 10

class BuggaloDialog(xbmcgui.WindowXMLDialog):
    THANK_YOU_VISIBLE_LABEL = 98
    DETAILS_VISIBLE_LABEL = 99
    CLOSE_BUTTON = 100
    SUBMIT_BUTTON = 101
    DETAILS_BUTTON = 102
    DETAILS_LIST = 103
    HEADING_LABEL = 110
    DETAILS_LINE1_LABEL = 111

    ERROR_MESSAGE_GROUP = 200
    DETAILS_GROUP = 201
    THANK_YOU_GROUP = 202
    THANK_YOU_LABEL = 203

    def __new__(cls, serviceUrl, gmailRecipient, heading, data):
        return super(BuggaloDialog, cls).__new__(cls, 'buggalo-dialog.xml', buggaloAddon.getAddonInfo('path'))

    def __init__(self, serviceUrl, gmailRecipient, heading, data):
        super(BuggaloDialog, self).__init__()
        self.serviceUrl = serviceUrl
        self.gmailRecipient = gmailRecipient
        self.heading = heading
        self.data = data
        self.detailsVisible = False


    def onInit(self):
        self.getControl(self.HEADING_LABEL).setLabel(self.heading)
        self.getControl(self.DETAILS_VISIBLE_LABEL).setVisible(not self.detailsVisible)
        self.getControl(self.THANK_YOU_VISIBLE_LABEL).setVisible(True)
        self.getControl(self.DETAILS_LINE1_LABEL).setLabel(buggaloAddon.getLocalizedString(91000) % self.data['addon']['name'].decode('utf-8', 'replace'))
        listControl = self.getControl(self.DETAILS_LIST)

        try:
            for group in sorted(self.data.keys()):
                values = self.data[group]
                if type(values) == dict:
                    item = xbmcgui.ListItem(label = '[B]%s[/B]' % group)
                    listControl.addItem(item)
                    keys = values.keys()
                    if group == 'userflow':
                        keys = sorted(keys)

                    for key in keys:
                            item = xbmcgui.ListItem(label = '    %s' % key, label2 = str(values[key]))
                            listControl.addItem(item)

                else:
                    item = xbmcgui.ListItem(label = '[B]%s[/B]' % group, label2 = str(values))
                    listControl.addItem(item)
        except Exception:
            item = xbmcgui.ListItem(label2 = buggaloAddon.getLocalizedString(91008))
            listControl.addItem(item)



    def onAction(self, action):
        if action.getId() in [ACTION_PARENT_DIR, ACTION_PREVIOUS_MENU]:
            self.close()

    def onClick(self, controlId):
        if controlId == self.CLOSE_BUTTON:
            self.close()

        elif controlId == self.SUBMIT_BUTTON:
            self.getControl(self.DETAILS_VISIBLE_LABEL).setVisible(True)
            self.getControl(self.THANK_YOU_VISIBLE_LABEL).setVisible(False)

            try:
                if self.serviceUrl is not None:
                    client.submitData(self.serviceUrl, self.data)
                elif self.gmailRecipient is not None:
                    client.emailData(self.gmailRecipient, self.data)
            except Exception:
                self.getControl(self.THANK_YOU_LABEL).setLabel(buggaloAddon.getLocalizedString(91009))

            xbmc.sleep(2000)
            self.close()

        elif controlId == self.DETAILS_BUTTON:
            self.detailsVisible = not self.detailsVisible
            self.getControl(self.DETAILS_VISIBLE_LABEL).setVisible(not self.detailsVisible)
            if self.detailsVisible:
                self.getControl(self.DETAILS_BUTTON).setLabel(buggaloAddon.getLocalizedString(91007))
            else:
                self.getControl(self.DETAILS_BUTTON).setLabel(buggaloAddon.getLocalizedString(91006))


    def onFocus(self, control):
        pass

lib/buggalo_userflow.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
#
import datetime
import os
import simplejson

import xbmc
import xbmcaddon

BUGGALO_ADDON = xbmcaddon.Addon('script.module.buggalo')
ADDON = xbmcaddon.Addon()

def trackUserFlow(value):
    userFlow = load()
    key = datetime.datetime.now().isoformat()

    userFlow[key] = value
    save(userFlow)

def load():
    path = xbmc.translatePath(BUGGALO_ADDON.getAddonInfo('profile'))
    file = os.path.join(path, '%s.json' % ADDON.getAddonInfo('id'))

    if os.path.exists(file):
        userFlow = simplejson.load(open(file))
    else:
        userFlow = dict()
    return userFlow

def save(userFlow):
    path = xbmc.translatePath(BUGGALO_ADDON.getAddonInfo('profile'))
    if not os.path.exists(path):
        os.makedirs(path)
    file = os.path.join(path, '%s.json' % ADDON.getAddonInfo('id'))

    try:
        # remove entries older than 24 hours
        # we compare strings rather the datetimes (a little hackish though)
        # but datetime.datetime.strptime() often fail for no apparent reason
        # see http://forum.xbmc.org/showthread.php?tid=112916
        oneDayAgo = datetime.datetime.now() - datetime.timedelta(days = 1)
        oneDayAgoStr = oneDayAgo.isoformat()
        for dateStr in userFlow.keys():
            if dateStr < oneDayAgoStr:
                del userFlow[dateStr]

        simplejson.dump(userFlow, open(file, 'w'))
    except Exception:
        print "problem saving userflow json file"



lib/buggalo_client.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
#
import datetime
import os
import platform
import simplejson
import sys
import traceback
import urllib2
import smtplib
from email.mime.text import MIMEText
import xbmc
import xbmcaddon

import buggalo_userflow as userflow

def gatherData(type, value, tracebackInfo, extraData, globalExtraData):
    data = dict()
    data['version'] = 4
    data['timestamp'] = datetime.datetime.now().isoformat()

    exception = dict()
    exception['stacktrace'] = traceback.format_tb(tracebackInfo)
    exception['type'] = str(type)
    exception['value'] = str(value)
    data['exception'] = exception

    system = dict()
    try:
        if hasattr(os, 'uname'):
            # Works on recent unix flavors
            (sysname, nodename, release, version, machine) = os.uname()
        else:
            # Works on Windows (and others?)
            (sysname, nodename, release, version, machine, processor) = platform.uname()

        system['nodename'] = nodename
        system['sysname'] = sysname
        system['release'] = release
        system['version'] = version
        system['machine'] = machine
    except Exception, ex:
        system['sysname'] = sys.platform
        system['exception'] = str(ex)
    data['system'] = system

    addon = xbmcaddon.Addon()
    addonInfo = dict()
    addonInfo['id'] = addon.getAddonInfo('id')
    addonInfo['name'] = addon.getAddonInfo('name')
    addonInfo['version'] = addon.getAddonInfo('version')
    addonInfo['path'] = addon.getAddonInfo('path')
    addonInfo['profile'] = addon.getAddonInfo('profile')
    data['addon'] = addonInfo

    xbmcInfo = dict()
    xbmcInfo['buildVersion'] = xbmc.getInfoLabel('System.BuildVersion')
    xbmcInfo['buildDate'] = xbmc.getInfoLabel('System.BuildDate')
    xbmcInfo['skin'] = xbmc.getSkinDir()
    xbmcInfo['language'] = xbmc.getInfoLabel('System.Language')
    data['xbmc'] = xbmcInfo

    execution = dict()
    execution['python'] = sys.version
    execution['sys.argv'] = sys.argv
    data['execution'] = execution

    data['userflow'] = userflow.loadUserFlow()

    extraDataInfo = dict()
    try:

        for (key, value) in globalExtraData.items():
            if isinstance(extraData, str):
                extraDataInfo[key] = value.decode('utf-8', 'ignore')
            elif isinstance(extraData, unicode):
                extraDataInfo[key] = value
            else:
                extraDataInfo[key] = str(value)

        if isinstance(extraData, dict):
            for (key, value) in extraData.items():
                if isinstance(extraData, str):
                    extraDataInfo[key] = value.decode('utf-8', 'ignore')
                elif isinstance(extraData, unicode):
                    extraDataInfo[key] = value
                else:
                    extraDataInfo[key] = str(value)
        elif extraData is not None:
            extraDataInfo[''] = str(extraData)
    except Exception, ex:
        (type, value, tb) = sys.exc_info()
        traceback.print_exception(type, value, tb)
        extraDataInfo['exception'] = str(ex)
    data['extraData'] = extraDataInfo

    return data


def submitData(serviceUrl, data):
    for attempt in range(0, 3):
        try:
            json = simplejson.dumps(data)
            req = urllib2.Request(serviceUrl, json)
            req.add_header('Content-Type', 'text/json')
            u = urllib2.urlopen(req)
            u.read()
            u.close()
            break # success; no further attempts
        except Exception:
            pass # probably timeout; retry


def emailData(recipient, data):
    """

    @param recipient:
    @param data:
    @return:
    """

    # build html table with data
    body = '<table border="1">'
    for group in sorted(data.keys()):
        values = data[group]
        if type(values) == dict:
            body += '<tr><td colspan="2"><h2>%s</h2></td></tr>' % group.capitalize()
            keys = values.keys()
            if group == 'userflow':
                keys = sorted(keys)

            for key in keys:
                body += '<tr><td>%s</td>' % key
                if key == 'stacktrace':
                    body += '<td><pre>'
                    for item in values[key]:
                        body += item + '\n'
                    body += '</pre></td>'
                else:
                    body += '<td>%s</td>' % str(values[key])
                body += '</tr>'
        else:
            body += '<tr><td><h2>%s</h2></td><td>%s</td></tr>' % (group.capitalize(), str(values))
    body += '</table>'

    msg = MIMEText(body, 'html')
    msg['Subject'] = '[Buggalo][%s] %s' % (data['addon']['id'], data['exception']['value'])
    msg['From'] = 'Buggalo'
    msg['To'] = recipient
    msg['X-Mailer'] = 'Buggalo Exception Collector'

    smtp = smtplib.SMTP('gmail-smtp-in.l.google.com')
    smtp.sendmail(msg['From'], msg['To'], msg.as_string(9))
    smtp.quit()
